Snowflake.Data/Core/Session/SFSession.cs

```diff
@@ -279,62 +279,75 @@ internal void close()
             if (!IsEstablished()) return;
             logger.Debug($"Closing session with id: {sessionId}, user: {_user}, database: {database}, schema: {schema}, role: {role}, warehouse: {warehouse}, connection start timestamp: {_startTime}");
             stopHeartBeatForThisSession();
+            var closeSessionRequest = PrepareCloseSessionRequest();
+            PostCloseSession(closeSessionRequest, restRequester);
+            sessionToken = null;
+        }
 
-            // Send a close session request
-            var queryParams = new Dictionary<string, string>();
-            queryParams[RestParams.SF_QUERY_SESSION_DELETE] = "true";
-            queryParams[RestParams.SF_QUERY_REQUEST_ID] = Guid.NewGuid().ToString();
-            queryParams[RestParams.SF_QUERY_REQUEST_GUID] = Guid.NewGuid().ToString();
+        internal void CloseNonBlocking()
+        {
+            // Nothing to do if the session is not open
+            if (!IsEstablished()) return;
+            logger.Debug($"Closing session with id: {sessionId}, user: {_user}, database: {database}, schema: {schema}, role: {role}, warehouse: {warehouse}, connection start timestamp: {_startTime}");
+            stopHeartBeatForThisSession();
+            var closeSessionRequest = PrepareCloseSessionRequest();
+            Task.Run(() => PostCloseSession(closeSessionRequest, restRequester));

```

Comment:
This code will be called from destructor of SnowflakeDbConnection. I would then move it to SfSession Destroy and make it implement IDisposable so it is clear what's going on.
What's more important I'm worried what happens when app is closing, can we still safely rely on thread pool here.

Reply:
as we discussed worse case is that it will leave session open and server will remove it afterwards

Reply:
I could move it but it does not matter since SfSession is not publicly accessible so no one can access to the session directly. It is always referenced either by SnowflakeDbConnection or connection pool.

Refactor which you propose would not change anything for the case when application is being closed.